### DISCUSSION TITLE
Multithread listen in server

### DISCUSSION THREAD

**Discussion by vanhaodev** [alec1o/Netly#46] [Q&A (Question and Answer)] [answered]
I want to create a while loop in async method for every client. But i see this lib only for single while loop to listen of all client request. Thanks

**Comment by alec1o** [comment]
Multithreading? Netly is multithreading... that is, each client runs in "parallel"
- 1x thread per server
- 1x thread per connected client (clientside or serverside)
Netly 3 use ``ThreadPool`` for running client sockets in  "parallel"

Netly 4: 
- Client (client-side and server-side) - don't use ThreadPool/Task/Thread on clients it's all async and have big performance improved and very low cpu usage than Netly 3.
- Server - don't use ThreadPool/Task on server it's all block method using 1x thread and have big performance improved.

--- 

Hi, feel free to interact here, hope to solve your problem... but I'm confused, I can't understand what you mean!

> Can you give pseudocode with this feature example?

Write what is the Netly behavior you wish to achieve (even if it is not supported)?

```cs
while (...?)
{
    ...?
}
...?
```

**Comment by vanhaodev** [comment]
Like this. Just a example. I dont want client wait other's client task on server
```
void Test()
{
      Server.onNewClient((newClient) =>
      {
          Listen(newClient);
      });
}
public async void Listen(Client client)
{
       while (client.isConnecting)
      {
          var packet = client.ListenPacket();
         if (packet == null) continue;
         HandlePacket(packet);
      }
}

```
thanks u. I wrote for mmorpg, if u have anyway to handle pls help me, thanks u. This lib so amazing

**Reply by alec1o** [reply]
Huh ok it's possible.

**Reply by vanhaodev** [reply]
```server.On.Accept((client) => { Listen(client)});```
is it ok?

**Reply by alec1o** [reply]
Exactly :100:

but this look more good.

Netly 4.x ([CURRENTLY] Unstable and undergoing changes)
Branch: https://github.com/alec1o/Netly/tree/v3.1.0

```cs
server.On.Accept((client) => Listen(ref client));

void Listen(ref ITCP.Client|IUDP.Client client)
{
    ...?
}
```

Netly 3.x (Stable) - Nuget & Unity Asset Store
Branch: https://github.com/alec1o/Netly/tree/dev

```cs
server.OnEnter((client) => Listen(ref client));

void Listen(ref TcpClient|UdpClient client)
{
    ...?
}
```

**Comment by alec1o** [comment, marked as answer]
> #### v3.x

<details>
<summary>Example Code</summary>

- Your way
    ```cs
    public void Start()
    {
        var server = new TcpServer() | UdpServer();
        
        server.OnEnter((client) =>
        {
            Task.Run(Listen(client));
        });
    }
    
    public async void Listen(TcpClient|UdpClient client)
    {
        List<byte[]> packages = new List<byte[]>();
    
        // this client have internal thread to receive and send package
        // it means: don't have server overhead
        client.OnData(data => packages.Add(data));
    
        while (client.IsOpened)
        {
            if (packages.Count <= 0) continue;
            byte[] package = packages[0];
            packages.RemoveAt(0); // remove used package
            HandlePacket(package);
        }
    }
    ```
- My Way
    ```cs
    var server = new TcpServer() | UdpServer();
    
    server.OnEnter((client) =>
    {
        // wanna client loop... called once per client
        Task.Run(async () =>
        {
            while (client.IsOpened)
            {
                ...?
            }
        });
    
        // handle package on my way
        client.OnData(data => HandlePacket(data));
    });
    ```

</details>

or

> #### v4.x

<details>
<summary>Example Code</summary>

- Your way
    ```cs
    public void Start()
    {
        var server = new TCP.Server() | UDP.Server();
    
        server.On.Accept((client) =>
        {
            Task.Run(Listen(client));
        });
    }

    public async void Listen(TCP.Client|UDP.Client client)
    {
        List<byte[]> packages = new List<byte[]>();
    
        // this client have internal thread to receive and send package
        // it means: don't have server overhead
        client.On.Data(data => packages.Add(data));
    
        while (client.IsOpened)
        {
            if (packages.Count <= 0) continue;
            byte[] package = packages[0];
            packages.RemoveAt(0); // remove used package
            HandlePacket(package);
        }
    }
    ```
- My Way
    ```cs
    var server = new TCP.Server() | UDP.Server() | RUDP.Server();

    server.On.Accept((client) =>
    {
        // wanna client loop... called once per client
        Task.Run(async () =>
        {
            while (client.IsOpened)
            {
                ...?
            }
        });
    
        // handle package on my way
        client.On.Data(data => HandlePacket(data));
    });
    ```

</details>

Apologies for the delay! :smile: 
Test and give feedback if it's working... :1st_place_medal: 

(Maybe it's a bit complex, I would make more effective examples for your specific use case, but i don't know your scope, that's why i did a generic example, feel free to refine it to your use case). :woman_technologist:

**Reply by vanhaodev** [reply]
Thank u so much, I will have so many topic for this library :)))
